File: src/aico/commands/edit.py
import os
import shlex
import subprocess
import sys
import tempfile
from dataclasses import replace
from pathlib import Path

import typer

from aico.console import is_input_terminal
from aico.diffing.stream_processor import recompute_derived_content
from aico.exceptions import ExternalDependencyError
from aico.models import AssistantChatMessage
from aico.session_loader import load_session_and_resolve_indices


def edit(
    index: str,
    prompt: bool,
) -> None:
    session, pair_indices, resolved_pair_index = load_session_and_resolve_indices(index)

    message_type: str
    target_message_index: int
    if prompt:
        message_type = "prompt"
        target_message_index = pair_indices.user_index
    else:
        message_type = "response"
        target_message_index = pair_indices.assistant_index

    target_message = session.data.chat_history[target_message_index]
    original_content = target_message.content

    new_content: str

    if not is_input_terminal():
        # Scripted mode: d new content from stdin
        new_content = sys.stdin.read()
    else:
        # Interactive mode: open editor
        fd, temp_file_path_str = tempfile.mkstemp(suffix=".md", text=True)
        temp_file_path = Path(temp_file_path_str)
        try:
            with os.fdopen(fd, "w") as f:
                _ = f.write(original_content)

            editor_cmd_str = os.environ.get("EDITOR", "vi")
            editor_cmd_parts = shlex.split(editor_cmd_str)
            full_command = editor_cmd_parts + [str(temp_file_path)]

            try:
                proc = subprocess.run(full_command, check=False)
            except FileNotFoundError:
                raise ExternalDependencyError(
                    f"Editor command not found: '{editor_cmd_parts[0]}'. "
                    + "Please set the $EDITOR environment variable."
                ) from None

            if proc.returncode != 0:
                raise ExternalDependencyError("Editor closed with non-zero exit code. Aborting.")

            new_content = temp_file_path.read_text()
        finally:
            temp_file_path.unlink(missing_ok=True)

    if new_content == original_content or not new_content:
        print("No changes detected. Aborting.")
        raise typer.Exit(code=0)

    updated_message = replace(target_message, content=new_content)
    new_asst_metadata: AssistantChatMessage | None = None

    # Invalidate derived content if editing an assistant response
    if isinstance(updated_message, AssistantChatMessage):
        new_derived_content = recompute_derived_content(
            assistant_content=new_content,
            context_files=session.data.context_files,
            session_root=session.root,
        )
        updated_message = replace(updated_message, derived=new_derived_content)
        new_asst_metadata = updated_message

    session.persistence.edit_message(target_message_index, new_content, new_asst_metadata)

    print(f"Updated {message_type} for message pair {resolved_pair_index}.")

File: src/commands/edit.rs
use crate::exceptions::AicoError;
use crate::session::Session;
use std::env;
use std::fs;
use std::io::IsTerminal;
use std::io::{Read, Write};
use std::process::Command;

fn run_editor(content: &str) -> Result<String, AicoError> {
    // 1. Create temp file
    let mut temp = tempfile::Builder::new().suffix(".md").tempfile()?;
    temp.write_all(content.as_bytes())?;
    let temp_path = temp.path().to_path_buf();

    // 2. Open Editor
    let editor = env::var("EDITOR").unwrap_or_else(|_| "vi".into());
    let parts = shlex::split(&editor).ok_or_else(|| {
        AicoError::Configuration(format!("Failed to parse EDITOR variable: '{}'", editor))
    })?;

    if parts.is_empty() {
        return Err(AicoError::Configuration(
            "EDITOR environment variable is empty".into(),
        ));
    }

    let status = Command::new(&parts[0])
        .args(&parts[1..])
        .arg(&temp_path)
        .status()
        .map_err(|e| {
            if e.kind() == std::io::ErrorKind::NotFound {
                AicoError::InvalidInput(format!(
                    "Editor '{}' not found. Please set $EDITOR.",
                    parts[0]
                ))
            } else {
                AicoError::Io(e)
            }
        })?;

    if !status.success() {
        return Err(AicoError::InvalidInput(format!(
            "Editor closed with exit code {}. Aborting.",
            status.code().unwrap_or(1)
        )));
    }

    // 3. Read back
    let mut buffer = String::new();
    fs::File::open(&temp_path)?.read_to_string(&mut buffer)?;
    Ok(buffer)
}

pub fn run(index_str: String, prompt_flag: bool) -> Result<(), AicoError> {
    let mut session = Session::load_active()?;
    let pair_idx = session.resolve_pair_index(&index_str)?;

    // Calculate which message within the pair to edit
    let msg_idx = if prompt_flag {
        pair_idx * 2
    } else {
        pair_idx * 2 + 1
    };

    let global_idx = session.view.message_indices[msg_idx];
    let records = session.store.read_many(&[global_idx])?;
    let record = records
        .first()
        .ok_or_else(|| AicoError::SessionIntegrity("Record not found".into()))?;

    let original_content = &record.content;

    // Detect if we have piped input. We only read from stdin if it's not a terminal.
    // To prevent hanging in tests or non-interactive environments where stdin is open but empty,
    // we use a check (if possible) or ensure our tests always provide EOF.
    // For Rust, the most reliable way to check for 'piped' input that doesn't hang
    // is often specific to the OS, but we can refine the logic to check if stdin is a terminal.

    let is_piped = !std::io::stdin().is_terminal();
    let force_editor = std::env::var("AICO_FORCE_EDITOR").is_ok();

    let new_content = if is_piped && !force_editor {
        let mut buffer = String::new();
        std::io::stdin().read_to_string(&mut buffer)?;
        buffer
    } else {
        run_editor(original_content)?
    };

    // Normalize for comparison
    let norm_new = new_content.replace("\r\n", "\n");
    let norm_old = original_content.replace("\r\n", "\n");

    if (norm_new.trim().is_empty() && is_piped && !force_editor) || norm_new == norm_old {
        println!("No changes detected. Aborting.");
        return Ok(());
    }

    // 4. Update session
    session.edit_message(msg_idx, norm_new)?;

    let target = if prompt_flag { "prompt" } else { "response" };
    println!("Updated {} for message pair {}.", target, pair_idx);

    Ok(())
}
